distributed/comm/mpi.py

```diff
@@ -0,0 +1,382 @@
+import asyncio  # for sleep
+import logging
+import threading
+
+from .registry import Backend, backends
+from .addressing import parse_host_port, unparse_host_port
+from .core import Comm, Connector, Listener, CommClosedError
+
+from ..protocol import nested_deserialize
+from ..utils import get_ip, get_ipv6, ensure_ip
+
+from tornado.ioloop import IOLoop
+
+
+# MPI communication layer.
+#
+# Host addresses are of the form "mpi://host ip/rank", e.g. "mpi:/10.0.0.10/4",
+# where rank is the MPI rank.  The host IPs are not required for MPI, but their
+# use is consistent with the other communication layers and can help with
+# debugging.
+#
+# The default project logging is overridden in this file such that each MPI
+# rank logs to a separate file, e.g. "rank2.log" for rank 2, which is useful for
+# debugging the setup and teardown of the communication layer.  But it is slow.
+
+
+# MPI communicator and rank.
+from mpi4py import MPI
+
+_mpi_comm = MPI.COMM_WORLD
+_mpi_rank = _mpi_comm.Get_rank()
+
+
+# logger = logging.getLogger(__name__)
+
+logging.basicConfig(
+    filename=f"rank{_mpi_rank}.log",
+    filemode="w",
+    level=logging.DEBUG,
+    format="%(asctime)s.%(msecs)03d %(levelname)s %(message)s",
+    datefmt="%H:%M:%S",
+)
+logger = logging.getLogger("mpi")
+
+
+# To identify the completion of asynchronous MPI sends and receives, need to
+# keep polling them.  This is the (minimum) time between polls, in seconds.
+_polling_time = 0.005
+
+
+class AsyncMPIRecv:
+    
+    def __init__(self, source, tag):
+        # source of None means receive from any rank.
+        self._source = source if source is not None else MPI.ANY_SOURCE
+        self._tag = tag
+        self._cancel = False
+        logger.debug(f"AsyncMPIRecv.__init__ source={self._source} tag={tag}")
+
+    def __await__(self):
+        while True:
+            if self._cancel:
+                return None, None
+            if _mpi_comm.iprobe(source=self._source, tag=self._tag):
```

Comment:
```python
            status = MPI.Status()
            if _mpi_comm.iprobe(source=self._source, tag=self._tag, status=status):
                source = status.Get_source()
                tag = status.Get_tag()
                msg = _mpi_comm.recv(source=source, tag=tag, status=status)
                return msg, status.Get_source()
```

Reply:
Alternatively, if `MPI.VERSION >= 3`, you could use the following code, which is the [thread-safe way](https://www.mpi-forum.org/docs/mpi-3.1/mpi31-report/node70.htm) of doing things:
```python
message = _mpi_comm.iprobe(source=self._source, tag=self._tag, status=status)
if message is not None:
    msg = message.recv(status=status)
    return msg, status.Get_source()
```